prow.sh

```diff
@@ -0,0 +1,667 @@
+# the repo supports "go test ./test/e2e -args --storage.testdriver" (https://github.com/kubernetes/kubernetes/pull/72836)
+# after setting KUBECONFIG. As a special case, if the repository is Kubernetes,
+# then `make WHAT=test/e2e/e2e.test` is called first to ensure that
+# all generated files are present.
+#
+# CSI_PROW_E2E_REPO=none disables E2E testing.
+#
+# TODO (?): support also an E2E test suite in the current repository.
+# TODO: upstream Kubernetes
+configvar CSI_PROW_E2E_REPO https://github.com/pohly/kubernetes "E2E repo"
+configvar CSI_PROW_E2E_VERSION storage-external-snapshot "E2E version"
+configvar CSI_PROW_E2E_IMPORT_PATH k8s.io/kubernetes "E2E package"
+
+# These control which tests run resp. don't run under "ginkgo -p".
+# Tests that depend on alpha features are disabled by default because
+# all tests with a feature tag are skipped.
+#
+# Note that feature tags in the test are set based on what is an alpha
+# feature in the *current* Kubernetes. When updating the test via
+# the CSI_PROW_E2E variables above and testing against an older Kubernetes
+# release, some tests might be enabled by default which don't pass
+# for that older Kubernetes version. In that case, the CSI_PROW_E2E_PARALLEL_SKIP_<k8s-version>
+# variable must be extended. <k8s-version> is just major+minor version
+# separated by underscore.
+configvar CSI_PROW_E2E_PARALLEL_FOCUS 'External.Storage.*csi-hostpath' "parallel E2E focus"
+# Raw block was an alpha feature in 1.13.
+configvar CSI_PROW_E2E_PARALLEL_SKIP_1_13 '|\[Testpattern: Dynamic PV .block volmode.\]' "additional parallel test skip for Kubernetes 1.13"
+configvar CSI_PROW_E2E_PARALLEL_SKIP "\\[Feature:|\\[Disruptive\\]$(eval echo "\${CSI_PROW_E2E_PARALLEL_SKIP_${csi_prow_kubernetes_version_suffix}}")" "parallel E2E skip"
+
+# After the parallel E2E test without alpha features, a test cluster
+# with alpha features is brought up and tests that were previously
+# disabled are run. This has to be enabled explicitly for each
+# Kubernetes release because it depends on the release whether there
+# are any alpha feature tests at all, which feature gates they depend
+# on and what the feature tags are.
+#
+# This second test run is sequential, therefore it can also be used to
+# enable disruptive tests. The gates can be empty for that. TODO (?):
+# also run tests sequentially without alpha features?
+#
+# If empty or "none", this phase is skipped.
+#
+# Testing against "latest" Kubernetes is problematic because some alpha
+# feature which used to work might stop working or change their behavior
+# such that the current tests no longer pass. If that happens,
+# kubernetes-csi components must be updated, either by disabling
+# the failing test for "latest" or by updating the test and not running
+# it anymore for older releases.
+configvar CSI_PROW_E2E_ALPHA_GATES_1_13 'VolumeSnapshotDataSource=true,BlockVolume=true,CSIBlockVolume=true' "alpha feature gates for Kubernetes 1.13"
```

Comment:
For regular CI Jobs, all of these config params are set in the .yaml in test-infra job config.  It seems easier to extend/add more jobs that way.

For example, if we want to add a 1.14 job in the future, then we would need to modify the config here, and then sync it to all the sidecars.

Reply:
If we start adding too many settings in the test-infra .yaml files, then we end up with a tight coupling between Prow job settings and what is getting tested. We may need different Prow jobs settings for different components. I think this will become complicated and fragile quickly.

The overall design is that the prow job just communicates the intent ("test against Kubernetes 1.14") and the component itself defines what that means.

Here's how testing against 1.14 can be rolled out:
- define the new Prow job without running it automatically
- update the components, verify that they pass the new Prow job by triggering it manually
- once all components are done, activate the Prow job

Reply:
I don't think we need to have custom focuses/skips for K8s e2es per sidecar.  The goal is to have uniform testing across all our sidecars and be able to test all the interaction together.

Right now with this current approach, to add a 1.14 job, we need to:
1. Add a new optional prow job in test-infra
2. Update the config here
3. Update the release-tools in every single sidecar
4. Test the job in all the sidecars
5. Enable the prow job by default in test-infra

If we define the parameters in the prow config, then the number of prs and steps are reduced:
1. Add the new optional prow jobs to test-infra
2. Test the jobs in all the sidecars (or just enable the CI jobs immediately and see which sidecars fail)
3. Enable the prow job by default

Reply:
In order to test the new PR job, you first need to have a pending PR in all the sidecars. For the less active repos there might not be one, so one would first have to create a fake one. Or can Prow run  the test against an already merged PR? No, doesn't look like it does although it theoretically could (https://github.com/kubernetes-csi/csi-driver-host-path/pull/26#issuecomment-475531854).

That's still a lot of manual work. In my proposal, creating the PR can be automated (https://github.com/kubernetes-csi/csi-release-tools/issues/7). My colleague @avalluri is currently looking into that.

Reply:
You can enable CI jobs on the sidecars.  In the ideal case they should pass because we're already testing against canary.  In the worst case, we'll know if something fails and we can make fixes in the appropriate repos.  Then once it passes, we can enable pull jobs.

Reply:
As discussed on Monday in the CSI standup meeting, we can defer this decision until later, because all settings inside `csi-release-tools` are just defaults. If it turns out to be desirable to have all settings in `test-infra`, then that can be done by setting env variables in the Prow job spec.

In the meantime I'll continue with the current approach, because it is more suitable for rapid prototyping (no need to change the job spec to try out something new).